lib/config/migrations.go

```diff
@@ -112,7 +114,25 @@ func migrateToConfigV38(cfg *Configuration) {
 		cfg.GUI.WebauthnRpId = defaultGuiCfg.WebauthnRpId
 	}
 	if cfg.GUI.WebauthnOrigin == "" {
-		cfg.GUI.WebauthnOrigin = defaultGuiCfg.WebauthnOrigin
+		_, port, err := net.SplitHostPort(cfg.GUI.RawAddress)
+		if err == nil {
+			scheme := "https"
+			if !cfg.GUI.RawUseTLS {
+				scheme = "http"
```

Comment:
This will only work if it's also opened via `localhost`, right?  I don't think the `UseTLS` option had any meaning in this context.

Reply:
Yeah. The fact that HTTPS is always available even when not forced actually makes this quite a lot more complicated. I had hoped we could get away with only a single allowed origin for starters - because that integrates much more easily with struct-field default tags and the config GUIs - but it seems like we do need to allow both `https://localhost:8384` and `http://localhost:8384` by default if we're dropping the HTTPS requirement. I'm looking into fixing this (the go-webauthn library already supports it), but it looks like it means we'll have to drop the `default` tag from `GUIConfiguration` since `SetDefaults` doesn't process defaults on slice-type fields.

Reply:
Fixed in 844516eca2735051a3faec04c529f35c4e29d08f. I refactored the config so that multiple origins can be allowed. The migration now adds both an `https://` and an `http://` origin if `cfg.GUI.RawUseTLS == false` at the time the migration runs, and adds only the `https://` origin if it is set to `true. I was half aware that the advanced settings GUI might already support slice fields using a comma-separated text input field, so it wasn't as bad as I worried it might be.

Reply:
This sound sensible.  Just to clarify once more:  I thought you said that HTTP is *only* considered a "secure location" (and thus available for WebAuthn) on localhost, right?  So we shouldn't be adding any settings if the currently used GUI address is not pointing to localhost.

On a related note, what about the loopback IPv4/v6 literals instead of the name localhost?  The default config uses `127.0.0.1:8384` IIRC.

Reply:
> you said that HTTP is _only_ considered a "secure location" (and thus available for WebAuthn) on localhost, right?

Yes, as far as I understand.

> On a related note, what about the loopback IPv4/v6 literals instead of the name localhost? The default config uses `127.0.0.1:8384` IIRC.

The `WebauthnOrigins` setting doesn't care about what the listen address setting is, only what's in the browser's address bar when you load the page. So since `localhost` (usually) resolves to `127.0.0.1`, that works fine as long as you open `http://localhost:8384` in the browser and not `http://127.0.0.1:8384` (and this is what the contextual "reload page at..." buttons do).

>So we shouldn't be adding any settings if the currently used GUI address is not pointing to localhost.

Hmyeah, sounds reasonable. But I'm not entirely sure whether there is an exhaustive list of listen addresses that `localhost` could successfully resolve to... is there?

Reply:
Sure, the setting is independent from the actual listen address.  And that's fine.  Just that we're trying to be extra smart in finding the default value now, and I think that should be as simple as possible:

1. Always initialize to a HTTPS value only during config migration.
2. If the GUI listen address points to a host name, use that.
3. Otherwise, use `localhost`.
4. Copy the port from the GUI listen address settings in any case.

Notice how `useTLS` plays no role in this.  If someone wants to use WebAuthn over plain HTTP, that should be configured manually (which is now possible with the slice of origins).  We always offer HTTPS, so we can be restrictive and simply rely on the redirection button if the automatic redirect is not enabled.  But none of that affects the setting's default value.

Reply:
Ok, should we remove the `if !c.UseTLS()` case from `defaultWebauthnOrigins()` then?
https://github.com/syncthing/syncthing/blob/fe68527f0a128ef92021fcd848a0ed9415f9e259/lib/config/guiconfiguration.go#L187-L193

> 2. If the GUI listen address points to a host name, use that.

Fixed in efd935a12bb6450ec1afd3671cdf8f4a79b5dd32.

Reply:
Ah - I got to your other comment on `migrations.go` now, which I'll take as a yes to the above question: https://github.com/syncthing/syncthing/pull/9175#discussion_r1730388276

Reply:
Fixed in d98703b09dcdf464076a087cae76c4a1960f59ea.